models/sla_policy/reply_time/int_zendesk__reply_time_combined.sql

```diff
@@ -110,45 +110,46 @@ with reply_time_calendar_hours_sla as (
-                or (agent_reply_at is null and next_solved_at is null)
-      then true
-      else false
-        end as is_sla_breached
-  from filtered_reply_times
-  
-), reply_time_breach as (
+), filtered_reply_times as ( 
   select 
-    *,
-    case when {{ dbt.datediff("sla_schedule_start_at", "agent_reply_at", 'minute') }} < 0 
-      then 0 
-      else sum_lapsed_business_minutes_new + {{ dbt.datediff("sla_schedule_start_at", "coalesce(agent_reply_at, current_time_check)", 'minute') }} 
-    end as sla_elapsed_time
-  from reply_time_breached_at_remove_old_sla
-)
+    * 
+  from lagging_time_block 
+  where ( 
+    in_business_hours 
+      and (( 
+        agent_reply_at >= sla_schedule_start_at and agent_reply_at <= sla_schedule_end_at) -- ticket is replied to between a schedule window 
+        or (agent_reply_at < sla_schedule_start_at and sum_lapsed_business_minutes_new = 0 and sla_breach_at = first_sla_breach_at) -- ticket is replied to before a schedule window and no business minutes have been spent on it 
+        or (agent_reply_at is null and {{ dbt.current_timestamp() }} >= sla_schedule_start_at and ({{ dbt.current_timestamp() }} < next_schedule_start or next_schedule_start is null)) -- ticket is not replied to and therefore active. But only bring through the active SLA record that is most recent (after the last SLA schedule starts but before the next, or if there does not exist a next SLA schedule start time)   
+        or (agent_reply_at >= sla_schedule_end_at and agent_reply_at <= next_schedule_start ) -- ticket is replied to outside sla schedule hours 
+      )) 
+    or not in_business_hours 
+ 
+), reply_time_breached_at_remove_old_sla as ( 
+  select 
+    *, 
+    {{ dbt.current_timestamp() }} as current_time_check, 
+    lead(sla_applied_at) over (partition by ticket_id, metric, in_business_hours order by sla_applied_at) as updated_sla_policy_starts_at, 
+    case when  
+      lead(sla_applied_at) over (partition by ticket_id, metric, in_business_hours order by sla_applied_at) --updated sla policy start at time 
+      < sla_breach_at then true else false end as is_stale_sla_policy, 
+    case when (sla_breach_at < agent_reply_at and sla_breach_at < next_solved_at) 
+                or (sla_breach_at < agent_reply_at and next_solved_at is null) 
+                or (agent_reply_at is null and sla_breach_at < next_solved_at) 
+                or (agent_reply_at is null and next_solved_at is null) 
+      then true 
+      else false 
+        end as is_sla_breached 
+  from filtered_reply_times 
+   
+), reply_time_breach as ( 
+  select  
+    *, 
+    case when {{ dbt.datediff("sla_schedule_start_at", "agent_reply_at", 'minute') }} < 0  
+      then 0  
+      else sum_lapsed_business_minutes_new + {{ dbt.datediff("sla_schedule_start_at", "coalesce(agent_reply_at, current_time_check)", 'minute') }}  
+    end as sla_elapsed_time 
+  from reply_time_breached_at_remove_old_sla 
+) 
```

Comment:
Thanks Joe, I worked through this logic and it makes sense to me. Also, definitely think it's clearer to break out the calculations and label them as their own fields like what you did with `runtime_minutes`, so I kept that in, though I moved some fields around. I also added some comments to explain the logic for future reference.

Using this query I compared times elapsed across ticket_metrics and sla_policies:

> 
> with sla_policies as 
>   (select ticket_id, metric, sla_elapsed_time from  `dbt-package-testing.zz_dbt_renee_zendesk.zendesk__sla_policies` 
> where metric = 'first_reply_time' ),
> 
> ticket_metrics_filtered as (
>   select ticket_id, first_reply_time_business_minutes from `dbt-package-testing.zz_dbt_renee_zendesk.zendesk__ticket_metrics` 
>   where ticket_id in (select ticket_id from sla_policies)
> )
> 
> select sla_policies.ticket_id, sla_policies.sla_elapsed_time, ticket_metrics_filtered.first_reply_time_business_minutes
> from sla_policies
> left join ticket_metrics_filtered
> on sla_policies.ticket_id = ticket_metrics_filtered.ticket_id

It looks mostly the same except 1. single minute differences and 2. I'll need to figure out why 80&81 aren't showing up in ticket_metrics

![image](https://github.com/fivetran/dbt_zendesk/assets/91097070/2a032d82-2a5c-430e-9220-b3ec36d56fee)

Reply:
Hi @fivetran-joemarkiewicz , following up on the above.

1. Your callout was right that for 80&81 agent_reply_at is null which explains why first_reply_time_business_minutes is null.
2.  I was able to investigate the minute discrepancies. My hunch was since it's never more than a minute, it had something to do with rounding the seconds in the two timestamps. I traced back the times for the relevant tickets and indeed `sla_elapsed_time` derives from timestamps involving seconds, whereas `first_reply_time_business_minutes` derives from fields that get calculated from # integer minutes from start of week, thus they don't involve seconds and I assume those # minutes were rounded amounts, rather than truncated amounts, based on what I tested. 

For example ticket 70
70: `schedule_start` = 2023-10-06 21:25:07 UTC, `agent_reply_at` = 2023-10-06 21:38:30 UTC = 13
This difference is 13 minutes. But once rounded (so 21:25:07 becomes 21:25 and 21:38:30 becomes 21:39), that difference is 14 minutes. This was corroborated by the other tickets (63, 68, 69)

For next steps, I know a minute difference isn't much but I imagine it would be a concern once a user starts aggregating minutes. Perhaps we should consider rounding the timestamps used to calculate `sla_elapsed_time`

Reply:
I cleaned up my queries and put them in a hex journal [here](https://app.hex.tech/fivetran/hex/67c8756e-e8d0-4e9f-a094-6afe27a8ed59/draft/logic).

Reply:
Thanks for investigating and getting to the bottom of this @fivetran-reneeli. Based off your finding here, wouldn't it make more sense to **not** round the `sla_elapsed_time` so we may be as accurate as possible? In this case I would recommend we should remove whatever rounding we have in the `zendesk__ticket_metrics` model. Although I would recommend that being its own separate ticket as it I would imagine that is outside the scope of this ticket.

Let me know if I am thinking about this differently from what you found or recommend.

Reply:
Right I should note that what was first popped up in my head, that we should fix from the `zendesk__ticket_metrics` side instead. 

But I don't think we can remove any 'rounding' because the integer minutes are provided from Zendesk itself, hence why I suggested doing the rounding to match. I forgot to mention that

Reply:
Where do we see that the minutes for the `zendesk__ticket_metrics` model are provided by Zendesk itself?

If that is the case, I would prefer we keep the discrepancy as we are being more accurate within the `zendesk__sla_policies` model. I would rather we stay accurate than round up.

Reply:
Yeah, for example if you look at the schedule object in the raw schema, `start_time` and `end_time` are in minutes rather than timestamp. Also I recall wasn't that one of the strange idiosyncrasies of the Zendesk data, that these fields aren't timestamps?

I'm not sure I'd immediately say sla_policies is more _accurate_. For example, for ticket 69 `sla_elapsed_time` says 4 minutes. But if you take the difference, technically it's closer to 3 minutes.
`dbt.datediff("sla_schedule_start_at", "agent_reply_at", 'minute')`

`2023-10-06 20:39:58 UTC, 2023-10-06 20:43:11 UTC `

Reply:
@fivetran-reneeli thanks for the input and you really helped shed some light into the sla_policies not actually being more accurate. When looking at this ticket in particular (in addition to the others) I was confused as to why the sla_policies model is not more accurate, because by all accounts it should be 🤔.

I looked into this some more and believe I found the reason why. It turns out in the `int_zendesk__reply_time_combined` model when we calculate the [running total minutes](https://github.com/fivetran/dbt_zendesk/blob/7b29ae004aefe3e96281417cb1638588035b0091/models/sla_policy/reply_time/int_zendesk__reply_time_combined.sql#L111) we are doing a datediff using the `minute` mark. However, across the package we are using the `second` mark and then dividing by 60 to get a more accurate metric ([datediff using seconds in the agent work time sla metrics](https://github.com/fivetran/dbt_zendesk/blob/7b29ae004aefe3e96281417cb1638588035b0091/models/sla_policy/agent_work_time/int_zendesk__agent_work_time_business_hours.sql#L61-L70), [datediff using seconds in the requester wait time sla metrics](https://github.com/fivetran/dbt_zendesk/blob/7b29ae004aefe3e96281417cb1638588035b0091/models/sla_policy/requester_wait_time/int_zendesk__requester_wait_time_business_hours.sql#L61-L70), [using seconds in the ticket metrics model calculations](https://github.com/fivetran/dbt_zendesk/blob/7b29ae004aefe3e96281417cb1638588035b0091/models/reply_times/int_zendesk__ticket_first_reply_time_business.sql#L40-L51), etc.).

I tested this locally and updated the datediff in the `int_zendesk__reply_time_combined` model and found the results to be more accurate to the actual. They are still different from the ticket_metrics results (as expected), but more accurate and in line with the initial assumption.

![image](https://github.com/fivetran/dbt_zendesk/assets/74217849/047a47a4-19a6-4315-9fd4-3272037b2177)

What are your thoughts?

Reply:
Alternatively, we could make the datediff in seconds instead of minutes, then divide by 60 and round to next minute. But I don't think it's a big enough discrepancy, after all it's something that falls back on what the function does, as opposed to our inherent logic, so it's not incorrect either. I guess it's more a matter of nuance, whether we want to truncate or round. 

We should focus more on making sure the bigger issue of the filter applied is corrected.

Reply:
I agree that the bigger issue is the filter (which I believe you have solved for in this PR); however, it is still important to understand the nuances of how this metric is being calculated across the different models as they do seem to be different which could cause some confusion for customers.

Let's go ahead and keep the `minute` datediff with no rounding for this PR. However, I feel the better ideal experience for customers is for our package to do no rounding. We should give them the metric in as close to the raw form as possible, and then let them decide if they wish to round. This would be a breaking change in my opinion as multiple records would all of a sudden change. Therefore, would you be able to create a new FR for us to leverage the `seconds` then divide by 60 (with no rounding) for the next breaking change? That will allow us to move forward with this PR in the meantime.

Reply:
Hi @fivetran-joemarkiewicz my bad-- I didn't see your previous two messages before I added my prior comment so it ended up sounding redundant. But yes, it's because the datediff uses minutes. I agree it makes sense that 'accurate' would mean divvying down to seconds first before converting to minutes, and it should be a separate PR. I'll open a FR for it 👍